ts/src/profiler.ts

```diff
@@ -13,47 +13,335 @@
+      return this.config.backoffMillis;
+    }
+    try {
+      await this.profileAndUpload(prof);
+    } catch (err) {
+      this.logger.error(`Error collecting and uploading profile: ${err}`);
+    }
+    return 0;
+  }
+
+  /**
+   * Talks to profiler server, which hangs until server indicates
+   * job should be profiled and then indicates what type of profile should
+   * be collected.
+   *
+   * If any problem is encountered, an error will be thrown.
+   *
+   * @return - a RequestProfile specifying which type of profile should be
+   * collected and other information needed to collect and upload a profile of
+   * the specified type.
+   *
+   * TODO (issue #28): right now, this call could hang for up to an hour when
+   * this method is the only thing on the event loop, keeping the program open
+   * even when all work is done. Should expose the ability to cancel the http
+   * request made here, and then determine when to cancel this request.
+   *
+   * Public to allow for testing.
+   */
+  async requestProfile(): Promise<RequestProfile> {
+    const reqBody = {
+      deployment: this.deployment,
+      profileType: this.profileTypes,
+    };
+    const options = {
+      method: 'POST',
+      uri: '/profiles',
+      body: reqBody,
+      json: true,
+    };
+
+    const [body, response] = await this.request(options);
+    if (!hasHttpStatusCode(response)) {
+      throw new Error('Server response missing status information.');
+    }
+    if (isErrorResponseStatusCode(response.statusCode)) {
+      let message: number|string = response.statusCode;
+      // tslint:disable-next-line: no-any
+      if ((response as any).statusMessage) {
+        message = response.statusMessage;
+      }
+      throw new Error(message.toString());
+    }
+    if (!isRequestProfile(body)) {
+      throw new Error(`Profile not valid: ${body}`);
+    }
+    return body;
+  }
+
+  /**
+   * Collects a profile of the type specified by the profileType field of prof.
+   * If any problem is encountered, like a problem collecting or uploading the
+   * profile, an error will be thrown.
+   *
+   * Public to allow for testing.
+   */
+  async profileAndUpload(prof: RequestProfile): Promise<void> {
+    prof = await this.profile(prof);
+    prof.labels = this.profileLabels;
+
+    const options = {
+      method: 'PATCH',
+      uri: API + '/' + prof.name,
+      body: prof,
+      json: true,
+    };
+    const [body, response] = await this.request(options);
+    if (!hasHttpStatusCode(response)) {
+      throw new Error(
+          'Server response missing status information when attempting to upload profile.');
+    }
+    if (isErrorResponseStatusCode(response.statusCode)) {
+      let message: number|string = response.statusCode;
+      // tslint:disable-next-line: no-any
+      if ((response as any).statusMessage) {
+        message = response.statusMessage;
+      }
+      throw new Error(`Could not upload profile: ${message}`);
```

Comment:
Nit: full stop is missing. Can you go over other messages and check that all has the full stop?

Reply:
done.